### ISSUE TITLE
REQ: Enable handling proxy paths

### ISSUE THREAD

**Issue by serussell** [xBrowserSync/API#14]
It would be helpful if `index.js` would not assume that the application is anchored at the root path, and would instead use `window.location.href` to figure out how to build a correct path in case xBrowserSync is proxied behind a different path.

**Comment by nero120** [comment]
Try this @serussell: #09be000.

**Comment by serussell** [comment]
@nero120 09be000 does the trick, thanks!

**Comment by serussell** [comment]
This is closed, but one other question I'm struggling with: can you suggest why the Javascript would make unauthenticated requests for some assets?  The third column is the user part of the authentication.  This is regardless of browser (same result from Safari as Vivaldi).

```
173.18.185.236 - ser [19/Jul/2018:18:03:26 -0400] "GET /styles.css HTTP/2.0" 200 136859
173.18.185.236 - - [19/Jul/2018:18:03:27 -0400] "GET /icon-128.png HTTP/2.0" 401 17
173.18.185.236 - - [19/Jul/2018:18:03:27 -0400] "GET /info HTTP/2.0" 401 17
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /89e40751d0e7b7f69bd0e2dec3e98f18.svg HTTP/2.0" 200 8094
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /448c34a56d699c29117adc64c43affeb.woff2 HTTP/2.0" 200 18028
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /587de8ec039052f50e69c9654439b991.woff2 HTTP/2.0" 200 15332
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /e21bf4e6adbbcebeedb2d078d9dbeeca.woff2 HTTP/2.0" 200 16992
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /c074f8ef4aea2b67fa0ae380041dacdf.woff2 HTTP/2.0" 200 15252
173.18.185.236 - ser [19/Jul/2018:18:03:27 -0400] "GET /74b994ea3b7ba91744de224045d6b36e.svg HTTP/2.0" 200 48683
173.18.185.236 - - [19/Jul/2018:18:03:27 -0400] "GET /icon-128.png HTTP/2.0" 401 17
```

This log happens to be when I hit the page from the subdomain rather than the proxied path; the results are the same no matter how I get there.

**Comment by nero120** [comment]
@serussell the xBrowserSync client does not authenticate to the API, all requests are sent unauthenticated. Your log shows that some assets are being denied with a 401 unauthenticated error (icon-128.png) whilst others are returned without requiring auth (styles.css), why that is I couldn't tell you, it would be down to your server config.

The reason xBrowserSync does not use any auth is to forego the need to store passwords on the server and therefore have the user be required to trust the server. Since the user's data is encrypted client-side with an encryption key that no one else knows, the user doesn't need to care about the API server as long as it is up and running. It also means the server admin doesn't need to worry about user management or provide features like profile management or "forgot my password", keeping the service lean and keeping the amount of "visible" user data to an absolute minimum.

**Comment by serussell** [comment]
@nero120, thanks for the response.  The reason for not *requiring* authentication is clear.  It seems that protecting, or restricting access to, a server, is not in a supported case. Another way of saying this is that anybody running a server commits to running a *public* server? Since the server can't differentiate between users, I can't even give preferential storage to family members, right?

**Comment by nero120** [comment]
Yes that's correct, there's currently no support to apply different rules (e.g. larger max sync size) to certain users. At the moment it's kind of a _security through obscurity_ approach if you want to run a private service (i.e. don't tell anyone else the IP/hostname! 🤐 

However, auth might be something I look at in future. As I run a [nextcloud](https://github.com/nextcloud/) instance I quite like the idea of porting the xBrowserSync API to a nextcloud app, which would require some means of providing username and password in the app. Though I wouldn't look at this until missing browser data syncing has been implemented and Firefox and Opera are supported.

**Comment by serussell** [comment]
Auth capability is important, but it's not (IMO) the responsibility of the server. That's something easily, and probably better, handled at the proxy. However, if you want to support it, you can't avoid putting that code in the client. The browser plug-in works I would assume because it's using the browser for the connection, and the browser can handle BASIC protected services. It's only the mobile client that's preventing users like myself from running a protected service, and having it available on mobile.

**Comment by nero120** [comment]
Can you break down for me how you've set up your auth on the xbs service? As I understood it, if you've set up basic auth then the client would be challenged for username and password when making a request to the service - so if it's working for the browser web extension how are you doing that?

**Comment by serussell** [comment]
First, sorry about any abruptness in my posts... I've been responding on my phone. 

I have this running behind Caddy, on which I've set up simple BASIC authentication.

**Comment by nero120** [comment]
Ah yes of course, the browser will automatically prompt you for creds when receiving a 401/407 response. The mobile apps are running inside of a web view which (according to some googling) appears to swallow the 401 challenges, hence no pop up for entering username and password. Ok well that explains what's going on.

Could I ask you to create a [new issue in the App repo](https://github.com/xBrowserSync/App/issues/new) requesting support for basic auth? Once you've done that we can close this issue and track it there, also any other users who would also like to see that functionality can upvote it. 😄

**Comment by NachoParra** [comment]
Hi guys,

Sorry to answer so late, and probably all will be resovled, by I have found this tool two weeks ago and I started working with this. I have my app behind a nginx proxy and working on a subfolder without any problem. This is the code I'0m using on the proxy:

```
location /xbrowsersync/ {
    auth_basic off;
    rewrite ^/xbrowsersync(/.*)$ $1 break;
    proxy_pass  http://192.168.1.175:28080/;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_redirect    off;
}
```

I still have to test adding basic authorization as I have on all my other services (nextcloud, sonarr, lazylibrarian, etc.), I didn't get enough time to do so. When I have this working I will be able to anser
